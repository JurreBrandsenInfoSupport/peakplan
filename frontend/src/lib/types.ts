export interface ProjectSummary {
  id: number;
  title: string;
}

export interface ProjectDetails {
  id: number;
  title: string;
  description: string;
}