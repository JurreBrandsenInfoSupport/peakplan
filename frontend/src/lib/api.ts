import {
  ProjectDetails,
  ProjectSummary,
  TaskResponseData as TaskListData,
} from "./types";

const backendUrl = "http://localhost:3000";

export class ApiError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
    Object.setPrototypeOf(this, ApiError.prototype);
  }
}

function getApiHost() {
  if (typeof window !== "undefined") {
    return "/api";
  }

  return backendUrl;
}

async function processResponse<T>(response: Response): Promise<T> {
  if (response.status === 404) {
    throw new ApiError("Not found", 404);
  }

  if (!response.ok) {
    const errorText = await response.text();
    throw new ApiError(
      `Error: ${response.status} ${response.statusText} - ${errorText}`,
      response.status,
    );
  }

  return await response.json();
}

export async function fetchProject(id: string): Promise<ProjectDetails> {
  return fetch(`${getApiHost()}/projects/${id}`).then(
    processResponse<ProjectDetails>,
  );
}

export async function fetchProjects(): Promise<ProjectSummary[]> {
  return fetch(`${getApiHost()}/projects`).then(
    processResponse<ProjectSummary[]>,
  );
}

export async function fetchProjectTasks(
  projectId: number,
): Promise<TaskListData> {
  return fetch(`${getApiHost()}/projects/${projectId}/tasks`).then(
    processResponse<TaskListData>,
  );
}

export async function fetchInboxTasks() {
  return fetch(`${getApiHost()}/tasks/inbox`).then(
    processResponse<TaskListData>,
  );
}

export async function fetchTodaysTasks() {
  return fetch(`${getApiHost()}/tasks/today`).then(
    processResponse<TaskListData>,
  );
}

export async function fetchWeeklyTasks() {
  return fetch(`${getApiHost()}/tasks/this-week`).then(
    processResponse<TaskListData>,
  );
}

export async function createProject(title: string, description: string) {
  return await fetch(`${getApiHost()}/projects`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ title, description }),
  }).then(processResponse<ProjectDetails>);
}

export async function createProjectTask(
  projectId: number,
  data: { title: string; description: string; deadline?: Date },
) {
  return fetch(`${getApiHost()}/projects/${projectId}/tasks`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  }).then(processResponse);
}

export async function updateTaskStatus(taskId: number, done: boolean) {
  return fetch(`${getApiHost()}/tasks/${taskId}`, {
    method: "PATCH",
    body: JSON.stringify({ done }),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export async function createTask({
  title,
  description,
  deadline,
}: {
  title: string;
  description: string;
  deadline?: Date;
}) {
  return fetch(`${getApiHost()}/tasks`, {
    method: "POST",
    body: JSON.stringify({ title, description, deadline }),
    headers: {
      "Content-Type": "application/json",
    },
  });
}
