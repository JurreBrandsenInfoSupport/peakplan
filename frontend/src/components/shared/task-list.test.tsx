import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TaskList from './task-list';
import { Task } from '@/lib/types';

describe('TaskList', () => {
    // Common handlers and setup
    const handleMarkTaskCompleted = vi.fn();
    const handleMarkTaskPending = vi.fn();
    const handleCreateTask = vi.fn();

    const samplePendingTask = {
        id: 1,
        title: 'test',
        description: '',
        deadline: new Date(2025, 1, 1),
        done: false
    };

    const sampleCompletedTask = { ...samplePendingTask, done: true };

    // Helper function to render the component
    const renderTaskList = (pendingTasks: Task[], completedTasks: Task[]) => {
        return render(<TaskList
            completedTasks={completedTasks}
            pendingTasks={pendingTasks}
            onMarkTaskCompleted={handleMarkTaskCompleted}
            onMarkTaskPending={handleMarkTaskPending}
            onCreateTask={handleCreateTask} />);
    };

    it('Renders pending tasks correctly', () => {
        const pendingTasks = [samplePendingTask];
        const completedTasks: Task[] = [];

        const component = renderTaskList(pendingTasks, completedTasks);
        expect(component).toMatchSnapshot();
    });

    it('Renders completed tasks correctly', () => {
        const pendingTasks: Task[] = [];
        const completedTasks = [sampleCompletedTask];

        const component = renderTaskList(pendingTasks, completedTasks);
        expect(component).toMatchSnapshot();
    });

    it('Marks tasks as pending correctly', () => {
        const pendingTasks: Task[] = [];
        const completedTasks = [sampleCompletedTask];

        const component = renderTaskList(pendingTasks, completedTasks);
        fireEvent.click(component.getByRole('checkbox'));

        expect(handleMarkTaskPending).toBeCalled();
    });

    it('Marks tasks as completed correctly', () => {
        const pendingTasks = [samplePendingTask];
        const completedTasks: Task[] = [];

        const component = renderTaskList(pendingTasks, completedTasks);
        fireEvent.click(component.getByRole('checkbox'));

        expect(handleMarkTaskCompleted).toBeCalled();
    });

    it('Creates a new task when the form for it submitted', async () => {
        const pendingTasks: Task[] = [];
        const completedTasks: Task[] = [];

        const user = userEvent.setup();
        const component = renderTaskList(pendingTasks, completedTasks);

        // Expand the create task form first.
        const expandButton = component.getByText('Add new task');
        await user.click(expandButton);

        // Next, input the values into the form, and submit it.
        const titleInput = component.getByPlaceholderText('Task title');
        const submitButton = component.getByText('Create Task');

        await user.type(titleInput, 'Test task');
        await user.click(submitButton);

        expect(handleCreateTask).toBeCalledWith("Test task", "", undefined);
    })
});