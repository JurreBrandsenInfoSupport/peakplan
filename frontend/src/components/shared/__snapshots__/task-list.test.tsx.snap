// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskList > Renders completed tasks correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="py-4 text-center text-muted-foreground font-medium"
      >
        Hurray, no pending tasks
      </div>
      <div
        class="mt-4"
        data-slot="collapsible"
        data-state="closed"
      >
        <button
          aria-controls="radix-«r2»"
          aria-expanded="false"
          class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full flex flex-row"
          data-slot="collapsible-trigger"
          data-state="closed"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-plus"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
           Add new task
        </button>
        <div
          class="mt-4"
          data-slot="collapsible-content"
          data-state="closed"
          hidden=""
          id="radix-«r2»"
          style=""
        />
      </div>
      <div
        data-slot="collapsible"
        data-state="open"
      >
        <div
          aria-controls="radix-«r3»"
          aria-expanded="true"
          class="w-full flex flex-row"
          data-slot="collapsible-trigger"
          data-state="open"
          type="button"
        >
          <h2
            class="font-semibold text-xl"
          >
            Completed tasks
          </h2>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevrons-up-down ml-auto"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m7 15 5 5 5-5"
            />
            <path
              d="m7 9 5-5 5 5"
            />
          </svg>
        </div>
        <div
          class="pt-4"
          data-slot="collapsible-content"
          data-state="open"
          id="radix-«r3»"
          style=""
        >
          <div
            class="rounded-md border"
          >
            <div
              class="relative w-full overflow-x-auto"
              data-slot="table-container"
            >
              <table
                class="w-full caption-bottom text-sm"
                data-slot="table"
              >
                <thead
                  class="[&_tr]:border-b"
                  data-slot="table-header"
                >
                  <tr
                    class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                    data-slot="table-row"
                  >
                    <th
                      class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-16"
                      data-slot="table-head"
                    />
                    <th
                      class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                      data-slot="table-head"
                    >
                      Title
                    </th>
                    <th
                      class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-64"
                      data-slot="table-head"
                    >
                      Deadline
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="[&_tr:last-child]:border-0"
                  data-slot="table-body"
                >
                  <tr
                    class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                    data-slot="table-row"
                    data-state="false"
                  >
                    <td
                      class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                      data-slot="table-cell"
                    >
                      <button
                        aria-checked="true"
                        class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
                        data-slot="checkbox"
                        data-state="checked"
                        role="checkbox"
                        type="button"
                        value="on"
                      >
                        <span
                          class="flex items-center justify-center text-current transition-none"
                          data-slot="checkbox-indicator"
                          data-state="checked"
                          style="pointer-events: none;"
                        >
                          <svg
                            aria-hidden="true"
                            class="lucide lucide-check size-3.5"
                            fill="none"
                            height="24"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M20 6 9 17l-5-5"
                            />
                          </svg>
                        </span>
                      </button>
                    </td>
                    <td
                      class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                      data-slot="table-cell"
                    >
                      test
                    </td>
                    <td
                      class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                      data-slot="table-cell"
                    >
                      2/1/2025
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="py-4 text-center text-muted-foreground font-medium"
    >
      Hurray, no pending tasks
    </div>
    <div
      class="mt-4"
      data-slot="collapsible"
      data-state="closed"
    >
      <button
        aria-controls="radix-«r2»"
        aria-expanded="false"
        class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full flex flex-row"
        data-slot="collapsible-trigger"
        data-state="closed"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
         Add new task
      </button>
      <div
        class="mt-4"
        data-slot="collapsible-content"
        data-state="closed"
        hidden=""
        id="radix-«r2»"
        style=""
      />
    </div>
    <div
      data-slot="collapsible"
      data-state="open"
    >
      <div
        aria-controls="radix-«r3»"
        aria-expanded="true"
        class="w-full flex flex-row"
        data-slot="collapsible-trigger"
        data-state="open"
        type="button"
      >
        <h2
          class="font-semibold text-xl"
        >
          Completed tasks
        </h2>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevrons-up-down ml-auto"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 15 5 5 5-5"
          />
          <path
            d="m7 9 5-5 5 5"
          />
        </svg>
      </div>
      <div
        class="pt-4"
        data-slot="collapsible-content"
        data-state="open"
        id="radix-«r3»"
        style=""
      >
        <div
          class="rounded-md border"
        >
          <div
            class="relative w-full overflow-x-auto"
            data-slot="table-container"
          >
            <table
              class="w-full caption-bottom text-sm"
              data-slot="table"
            >
              <thead
                class="[&_tr]:border-b"
                data-slot="table-header"
              >
                <tr
                  class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                  data-slot="table-row"
                >
                  <th
                    class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-16"
                    data-slot="table-head"
                  />
                  <th
                    class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                    data-slot="table-head"
                  >
                    Title
                  </th>
                  <th
                    class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-64"
                    data-slot="table-head"
                  >
                    Deadline
                  </th>
                </tr>
              </thead>
              <tbody
                class="[&_tr:last-child]:border-0"
                data-slot="table-body"
              >
                <tr
                  class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                  data-slot="table-row"
                  data-state="false"
                >
                  <td
                    class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                    data-slot="table-cell"
                  >
                    <button
                      aria-checked="true"
                      class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
                      data-slot="checkbox"
                      data-state="checked"
                      role="checkbox"
                      type="button"
                      value="on"
                    >
                      <span
                        class="flex items-center justify-center text-current transition-none"
                        data-slot="checkbox-indicator"
                        data-state="checked"
                        style="pointer-events: none;"
                      >
                        <svg
                          aria-hidden="true"
                          class="lucide lucide-check size-3.5"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M20 6 9 17l-5-5"
                          />
                        </svg>
                      </span>
                    </button>
                  </td>
                  <td
                    class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                    data-slot="table-cell"
                  >
                    test
                  </td>
                  <td
                    class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                    data-slot="table-cell"
                  >
                    2/1/2025
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TaskList > Renders pending tasks correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-md border"
      >
        <div
          class="relative w-full overflow-x-auto"
          data-slot="table-container"
        >
          <table
            class="w-full caption-bottom text-sm"
            data-slot="table"
          >
            <thead
              class="[&_tr]:border-b"
              data-slot="table-header"
            >
              <tr
                class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                data-slot="table-row"
              >
                <th
                  class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-16"
                  data-slot="table-head"
                />
                <th
                  class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                  data-slot="table-head"
                >
                  Title
                </th>
                <th
                  class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-64"
                  data-slot="table-head"
                >
                  Deadline
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
              data-slot="table-body"
            >
              <tr
                class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
                data-slot="table-row"
                data-state="false"
              >
                <td
                  class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                  data-slot="table-cell"
                >
                  <button
                    aria-checked="false"
                    class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
                    data-slot="checkbox"
                    data-state="unchecked"
                    role="checkbox"
                    type="button"
                    value="on"
                  />
                </td>
                <td
                  class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                  data-slot="table-cell"
                >
                  test
                </td>
                <td
                  class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                  data-slot="table-cell"
                >
                  2/1/2025
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="mt-4"
        data-slot="collapsible"
        data-state="closed"
      >
        <button
          aria-controls="radix-«r0»"
          aria-expanded="false"
          class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full flex flex-row"
          data-slot="collapsible-trigger"
          data-state="closed"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-plus"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
           Add new task
        </button>
        <div
          class="mt-4"
          data-slot="collapsible-content"
          data-state="closed"
          hidden=""
          id="radix-«r0»"
          style=""
        />
      </div>
      <div
        data-slot="collapsible"
        data-state="closed"
      >
        <div
          aria-controls="radix-«r1»"
          aria-expanded="false"
          class="w-full flex flex-row"
          data-slot="collapsible-trigger"
          data-state="closed"
          type="button"
        >
          <h2
            class="font-semibold text-xl"
          >
            Completed tasks
          </h2>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevrons-up-down ml-auto"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m7 15 5 5 5-5"
            />
            <path
              d="m7 9 5-5 5 5"
            />
          </svg>
        </div>
        <div
          class="pt-4"
          data-slot="collapsible-content"
          data-state="closed"
          hidden=""
          id="radix-«r1»"
          style=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-md border"
    >
      <div
        class="relative w-full overflow-x-auto"
        data-slot="table-container"
      >
        <table
          class="w-full caption-bottom text-sm"
          data-slot="table"
        >
          <thead
            class="[&_tr]:border-b"
            data-slot="table-header"
          >
            <tr
              class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
              data-slot="table-row"
            >
              <th
                class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-16"
                data-slot="table-head"
              />
              <th
                class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                data-slot="table-head"
              >
                Title
              </th>
              <th
                class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-64"
                data-slot="table-head"
              >
                Deadline
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
            data-slot="table-body"
          >
            <tr
              class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
              data-slot="table-row"
              data-state="false"
            >
              <td
                class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                data-slot="table-cell"
              >
                <button
                  aria-checked="false"
                  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
                  data-slot="checkbox"
                  data-state="unchecked"
                  role="checkbox"
                  type="button"
                  value="on"
                />
              </td>
              <td
                class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                data-slot="table-cell"
              >
                test
              </td>
              <td
                class="p-2 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
                data-slot="table-cell"
              >
                2/1/2025
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="mt-4"
      data-slot="collapsible"
      data-state="closed"
    >
      <button
        aria-controls="radix-«r0»"
        aria-expanded="false"
        class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full flex flex-row"
        data-slot="collapsible-trigger"
        data-state="closed"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
         Add new task
      </button>
      <div
        class="mt-4"
        data-slot="collapsible-content"
        data-state="closed"
        hidden=""
        id="radix-«r0»"
        style=""
      />
    </div>
    <div
      data-slot="collapsible"
      data-state="closed"
    >
      <div
        aria-controls="radix-«r1»"
        aria-expanded="false"
        class="w-full flex flex-row"
        data-slot="collapsible-trigger"
        data-state="closed"
        type="button"
      >
        <h2
          class="font-semibold text-xl"
        >
          Completed tasks
        </h2>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevrons-up-down ml-auto"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 15 5 5 5-5"
          />
          <path
            d="m7 9 5-5 5 5"
          />
        </svg>
      </div>
      <div
        class="pt-4"
        data-slot="collapsible-content"
        data-state="closed"
        hidden=""
        id="radix-«r1»"
        style=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
